internal/parsers/hclparser/hcltype_test.go

```diff
@@ -1,270 +1,180 @@
-				Label:  "another_object",
+		{
+			expr: "list(list(list(object(foo))))",
+			want: entities.Type{
+				TFType: types.TerraformList,
+				Nested: &entities.Type{
+					TFType: types.TerraformList,
+					Nested: &entities.Type{
+						TFType: types.TerraformList,
+						Nested: &entities.Type{
+							TFType: types.TerraformObject,
+							Label:  "foo",
+						},
+					},
+				},
 			},
 		},
-	},
-	{
-		expression: `object(my_object_name)`,
-		want: entities.Type{
-			TFType: types.TerraformObject,
-			Label:  "my_object_name",
-		},
-	},
-	{
-		expression: `map(my_object_name)`,
-		want: entities.Type{
-			TFType: types.TerraformMap,
-			Nested: &entities.Type{
-				TFType: types.TerraformObject,
-				Label:  "my_object_name",
+		{
+			expr: "set(number)",
+			want: entities.Type{
+				TFType: types.TerraformSet,
+				Nested: &entities.Type{
+					TFType: types.TerraformNumber,
+				},
 			},
 		},
-	},
-	{
-		expression: `object(another_object_name)`,
-		want: entities.Type{
-			TFType: types.TerraformObject,
-			Label:  "another_object_name",
-		},
-	},
-	{
-		expression: `string`,
-		want: entities.Type{
-			TFType: types.TerraformString,
-		},
-	},
-	{
-		expression: `number`,
-		want: entities.Type{
-			TFType: types.TerraformNumber,
-		},
-	},
-	{
-		expression: `bool`,
-		want: entities.Type{
-			TFType: types.TerraformBool,
-		},
-	},
-	{
-		expression: `resource(foo_bar_baz)`,
-		want: entities.Type{
-			TFType: types.TerraformResource,
-			Label:  "foo_bar_baz",
+		{
+			expr: "set(string)",
+			want: entities.Type{
+				TFType: types.TerraformSet,
+				Nested: &entities.Type{
+					TFType: types.TerraformString,
+				},
+			},
 		},
-	},
-}
+	}
 
-func TestGetOutputTypeFromExpression(t *testing.T) {
-	for _, tt := range outputTests {
-		t.Run(tt.expression, func(t *testing.T) {
-			expr, parseDiags := hclsyntax.ParseExpression([]byte(tt.expression), "", hcl.Pos{Line: 1, Column: 1, Byte: 0})
+	for _, tt := range testCases {
+		t.Run(tt.expr, func(t *testing.T) {
+			hclExpr, parseDiags := hclsyntax.ParseExpression([]byte(tt.expr), "", hcl.Pos{Line: 1, Column: 1, Byte: 0})
 			if parseDiags.HasErrors() {
-				t.Errorf("Error parsing expression: %v", parseDiags.Errs())
+				t.Fatalf("Error parsing expression: %v", parseDiags.Errs())
 			}
 
-			got, err := GetOutputTypeFromExpression(expr)
-			assert.NoError(t, err)
+			got, err := processType(hclExpr, types.TerraformEmptyType)
+			if err != nil {
+				t.Logf("[%s] ERROR: %v", tt.expr, err)
+			}
+			assert.NoError(t, err, fmt.Sprintf("[ERR] evaluating expr %s", tt.expr))
 
-			test.AssertEqualTypes(t, tt.want, got)
+			assertEqualTypes(t, tt.want, got)
 		})
 	}
 }
+
+func assertEqualTypes(t *testing.T, want, got entities.Type) {
+	t.Helper()
+
+	assert.EqualStrings(t, want.TFType.String(), got.TFType.String())
+	assert.EqualStrings(t, want.Label, got.Label)
+
+	if want.Nested != nil && got.Nested != nil {
+		assertEqualTypes(t, *want.Nested, *got.Nested)
+	}
```

Comment:
if want.Nested != nil and got.Nested == nil nothing will be done (success), but I imagine in this case it should fail since you wanted nested and got nothing :thinking:

Reply:
You can go with something like:

```go
if (want.Nested == nil) != (got.Nested == nil) {
    t.Fatalf("want.Nested[%+v] != got.Nested[%+v]",
			want.Nested, got.Nested)
}

if want.Nested == nil {
    return //both are nil in this case, nothing to do
}

// For sure both are non-nil
assertEqualTypes(t, *want.Nested, *got.Nested)
```

Which checks first if both are nil OR both are non-nil. After that you can just return if want or got is nil. And finally compare because the last case is both are non-nil.